### PR TITLE
Terraform for Provision ECR and Lambda

### LINKED ISSUE #17: Terraform Scripts for the Provisioning Lambda
### Steps
- [x] Write Terraform scripts to deploy provisioning lambda
- [x] Include ECR integration and Lambda setup in the scripts
- [x] Test the Terraform scripts for correct deployment
- [x] Document the Terraform deployment process

### User Story
As a cloud engineer, I need Terraform scripts to automate the deployment of the provisioning lambda to ensure consistency and repeatability.

### PR DESCRIPTION
**Pull request by danfh00**
Creates the terraform files needed to make the provision lambda and the ECR it builds from.
Closes #17

### CODE REVIEW

File: ETL/Provision/aws_infrastructure/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# output "lambda_arn" {
+#   value = aws_lambda_function.provision_lambda.arn
+# }
```
Comment (BerkayDur) [inline ETL/Provision/aws_infrastructure/outputs.tf]:
```suggestion
 output "lambda_arn" {
   value = aws_lambda_function.provision_lambda.arn
 }
```

File: ETL/Provision/aws_infrastructure/main.tf
Comment (BerkayDur) [inline ETL/Provision/aws_infrastructure/main.tf:1]: Nice. Just need to add the Eventbridge scheduler and then it's all complete.

**Review by BerkayDur — CHANGES_REQUESTED** [review]

File: ETL/Provision/aws_infrastructure/main.tf [hunk 2]
```diff
@@ -0,0 +1,57 @@
+provider "aws" {
```
Comment (A-Faris) [inline ETL/Provision/aws_infrastructure/main.tf:3]:
```suggestion
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.0"
    }
  }
}

provider "aws" {
```
Could add but not needed. Found it here:
https://registry.terraform.io/providers/hashicorp/aws/latest/docs

File: ETL/Provision/aws_infrastructure/main.tf [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    region = var.AWS_REGION
+    access_key = var.ACCESS_KEY
+    secret_key = var.SECRET_ACCESS_KEY
```
Comment (A-Faris) [inline ETL/Provision/aws_infrastructure/main.tf]: Following the [AWS Configuration Reference](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#aws-configuration-reference) looks nice

File: ETL/Provision/aws_infrastructure/main.tf [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  })
+}
+
+resource "aws_ecr_repository" "provision-repo" {
+  name                 = "hermes-provision-lambda"
+  image_tag_mutability = "MUTABLE"
+  image_scanning_configuration {
+    scan_on_push = false
+  }
+}
```
Comment (A-Faris) [inline ETL/Provision/aws_infrastructure/main.tf]: Not needed as terraform doesn't allow to add images. Find the image Uri and use that.

File: ETL/Provision/aws_infrastructure/main.tf [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+}
+
+resource "aws_lambda_function" "provision_lambda" {
+  function_name = "c11-hermes-provision"
+  role = aws_iam_role.provision_lambda_role.arn
+  package_type = "Image"
+  image_uri = "129033205317.dkr.ecr.eu-west-2.amazonaws.com/c11-hermes-provision:latest"
+  architectures = ["x86_64"]
+  timeout       = 30
```
Comment (A-Faris) [inline ETL/Provision/aws_infrastructure/main.tf]:
```suggestion
resource "aws_lambda_function" "provision_lambda" {
  function_name = "c11-hermes-provision"
  role          = aws_iam_role.provision_lambda_role.arn
  package_type  = "Image"
  image_uri     = "129033205317.dkr.ecr.eu-west-2.amazonaws.com/c11-hermes-provision:latest"
  architectures = ["x86_64"]
  timeout       = 30
```
Fixing spacing makes it easier to read

File: ETL/Provision/aws_infrastructure/main.tf [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
```
Comment (A-Faris) [inline ETL/Provision/aws_infrastructure/main.tf:38]:
```suggestion
    logging_config {
        log_format        = "Text"
        log_group         = "/aws/lambda/c11-hermes-clean_lambda"
    }
```
Hopefully this adds logging

File: ETL/Provision/aws_infrastructure/main.tf (hunk 4 above)
Comment (A-Faris) [reply ETL/Provision/aws_infrastructure/main.tf]:
```suggestion

```

File: ETL/Provision/aws_infrastructure/main.tf (hunk 5 above)
Comment (A-Faris) [reply ETL/Provision/aws_infrastructure/main.tf]: Do this whenever you can

**Review by A-Faris — CHANGES_REQUESTED** [review]
Also add docstring on each file explaining what it does briefly

File: ETL/Provision/aws_infrastructure/outputs.tf (hunk 1 above)
Comment (danfh00) [reply ETL/Provision/aws_infrastructure/outputs.tf]: done

File: ETL/Provision/aws_infrastructure/main.tf (hunk 4 above)
Comment (danfh00) [reply ETL/Provision/aws_infrastructure/main.tf]: done

File: ETL/Provision/aws_infrastructure/main.tf (hunk 6 above)
Comment (danfh00) [reply ETL/Provision/aws_infrastructure/main.tf:38]: added

File: ETL/Provision/aws_infrastructure/main.tf
Comment (danfh00) [reply ETL/Provision/aws_infrastructure/main.tf:1]: That is for the step function, not the provision lambda😎

**Review by A-Faris — APPROVED** [review]
All good

**Review by BerkayDur — APPROVED** [review]
Seems good!

**Pull request merged**